Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,264 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.FiniteType
+import Mathlib.RingTheory.Nakayama
+import Mathlib.Algebra.Module.FinitePresentation
+import Mathlib.LinearAlgebra.TensorProduct.RightExactness
+import Mathlib.RingTheory.Flat.Basic
+import Mathlib.RingTheory.Ideal.LocalRing
+
+/-!
+# Finite modules over local rings
+
+This file gathers various results of finite modules over a local ring `(R, 𝔪, k)`.
+
+## Main results
+- `LocalRing.subsingleton_tensorProduct`: If `M` is finitely generated, `k ⊗ M = 0 ↔ M = 0`.
+- `Module.free_of_maximalIdeal_rTensor_injective`:
+  If `M` is a finitely presented module such that `m ⊗ M → M` is injective
+  (for example when `M` is flat), then `M` is free.
+- `Module.free_of_lTensor_residueField_injective`: If `N → M → P → 0` is a presentation of `P` with
+  `N` finite and `M` finite free, then `k ⊗ N → k ⊗ M` is injective implies `P` is free.
+- `LocalRing.split_injective_iff_lTensor_residueField_injective`:
+  Given an `R`-linear map `l : M → N` with `M` finite and `N` is finite free,
+  `l` is a split injection iff `k ⊗ l` is a (split) injection.
+-/
+
+
+
+variable {R S} [CommRing R] [CommRing S] [Algebra R S]
+
+section
+
+variable [LocalRing R] {M N} [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+
+open Function (Injective Surjective Exact)
+open LocalRing TensorProduct
+
+local notation "k" => ResidueField R
+local notation "𝔪" => maximalIdeal R
+
+variable {P} [AddCommGroup P] [Module R P] (f : M →ₗ[R] N) (g : N →ₗ[R] P)
+variable (hg : Surjective g) (h : Exact f g)
+
+theorem LocalRing.map_mkQ_eq_top {N : Submodule R M} [Module.Finite R M] :
+    N.map (Submodule.mkQ (𝔪 • ⊤)) = ⊤ ↔ N = ⊤ := by
+  constructor
```

Comment:
If you put `LocalRing.map_mkQ_eq_top` *after* `LocalRing.map_mkQ_eq`, its proof simply becomes
```
  convert LocalRing.map_mkQ_eq (N₁ := N) (N₂ := ⊤) le_top Module.Finite.out
  simp only [Submodule.map_top, Submodule.range_mkQ]
  ```